### PR TITLE
#307: Csrf starter

### PR DESCRIPTION
**Pull request by sujith-mn**

### CODE REVIEW

File: documentation/guide-csrf.asciidoc [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+----
+
+Also remove  .csrf.disable()  if exists.
+
+Configure environment.ts file to use the CSRF in the front end.
+
+[source]
+----
+security: 'csrf'
+----
```
Comment (hohwille) [inline documentation/guide-csrf.asciidoc:55]: IMHO we should better link directly to the `devon4ng` guide explaining this option. If that setting changes we will miss it in `devon4j` and the documentation is "broken".

File: modules/security-csrf/src/main/java/com/devonfw/module/security/csrf/common/api/NlsBundleApplicationRoot.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package com.devonfw.module.security.csrf.common.api;
+
+import net.sf.mmm.util.nls.api.NlsBundle;
+import net.sf.mmm.util.nls.api.NlsBundleMessage;
+import net.sf.mmm.util.nls.api.NlsMessage;
+
+/**
+ * This is the {@link NlsBundle} for this application.
+ */
+public interface NlsBundleApplicationRoot extends NlsBundle {
```
Comment (hohwille) [inline modules/security-csrf/src/main/java/com/devonfw/module/security/csrf/common/api/NlsBundleApplicationRoot.java:10]: please remove this class.

File: modules/security-csrf/pom.xml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  <artifactId>devon4j-security-csrf</artifactId>
+  <version>${devon4j.version}</version>
+  <packaging>jar</packaging>
+  <name>${project.artifactId}</name>
+  <description>Security Module of the Open Application Standard Platform for Java (devon4j) specifically for csrf.</description>
+
+  <dependencies>
+    <dependency>
+      <groupId>jakarta.transaction</groupId>
+    <artifactId>jakarta.transaction-api</artifactId>
```
Comment (hohwille) [inline modules/security-csrf/pom.xml:21]: indendation incorrect. Always use auto-format (`[Shift][Ctrl][f]`).

File: modules/security-csrf/src/main/java/com/devonfw/module/security/csrf/common/api/UserProfile.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package com.devonfw.module.security.csrf.common.api;
+
+public interface UserProfile {
```
Comment (hohwille) [inline modules/security-csrf/src/main/java/com/devonfw/module/security/csrf/common/api/UserProfile.java:3]: please remove this class. Not related to CSRF.

File: modules/security-csrf/src/main/java/com/devonfw/module/security/csrf/common/api/exception/ApplicationBusinessException.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
+package com.devonfw.module.security.csrf.common.api.exception;
+
+import net.sf.mmm.util.nls.api.NlsMessage;
+
+/**
+ * Abstract base class for business exceptions of this application.
+ */
+public abstract class ApplicationBusinessException extends ApplicationException {
```
Comment (hohwille) [inline modules/security-csrf/src/main/java/com/devonfw/module/security/csrf/common/api/exception/ApplicationBusinessException.java:8]: We removed all this code from app templates. It is not needed for CSRF so please remove.

File: modules/security-csrf/src/main/java/com/devonfw/module/security/csrf/common/api/exception/ApplicationException.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
+package com.devonfw.module.security.csrf.common.api.exception;
+
+import net.sf.mmm.util.exception.api.NlsRuntimeException;
+import net.sf.mmm.util.nls.api.NlsMessage;
+
+/**
+ * Abstract base class for custom exceptions of this application.
+ */
+public abstract class ApplicationException extends NlsRuntimeException {
```
Comment (hohwille) [inline modules/security-csrf/src/main/java/com/devonfw/module/security/csrf/common/api/exception/ApplicationException.java:9]: Please remove.

File: modules/security-csrf/src/main/java/com/devonfw/module/security/csrf/common/api/exception/NoActiveUserException.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
+package com.devonfw.module.security.csrf.common.api.exception;
+
+import com.devonfw.module.security.csrf.common.api.NlsBundleApplicationRoot;
+
+/**
+ * Thrown when an operation is requested that requires a user to be logged in, but no such user exists.
+ */
+public class NoActiveUserException extends ApplicationBusinessException {
```
Comment (hohwille) [inline modules/security-csrf/src/main/java/com/devonfw/module/security/csrf/common/api/exception/NoActiveUserException.java:8]: Please remove.

File: modules/security-csrf/src/main/java/com/devonfw/module/security/csrf/common/api/to/UserProfileTo.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
+package com.devonfw.module.security.csrf.common.api.to;
+
+import com.devonfw.module.basic.common.api.to.AbstractTo;
+import com.devonfw.module.security.csrf.common.api.UserProfile;
+
+/**
+ * Implementation of {@link UserProfile} as {AbstractTo TO}.
+ */
+public class UserProfileTo extends AbstractTo implements UserProfile {
```
Comment (hohwille) [inline modules/security-csrf/src/main/java/com/devonfw/module/security/csrf/common/api/to/UserProfileTo.java:9]: Please remove.

File: modules/security-csrf/src/main/java/com/devonfw/module/security/csrf/service/api/rest/CsrfRestService.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+import org.springframework.security.web.csrf.CsrfToken;
+
+import com.devonfw.module.rest.common.api.RestService;
+
+/**
+ * The CSRF security REST service provides access to the csrf token, the authenticated user's meta-data. Furthermore, it
+ * provides functionality to check permissions and roles of the authenticated user.
+ */
+@Path("/security/v1")
```
Comment (hohwille) [inline modules/security-csrf/src/main/java/com/devonfw/module/security/csrf/service/api/rest/CsrfRestService.java:17]: IMHO we should replace `security` with `csrf`. This is not the "universal generic security service" anymore and that is on purpose as we want to be modular and avoid mixing multiple features that may collide.

File: modules/security-csrf/src/main/java/com/devonfw/module/security/csrf/service/api/rest/CsrfRestService.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+public interface CsrfRestService extends RestService {
+
+  /**
+   * @param request {@link HttpServletRequest} to retrieve the current session from.
+   * @param response {@link HttpServletResponse} to send additional information.
+   * @return the Spring Security {@link CsrfToken} from the server session.
+   */
+  @GET
+  @Path("/csrftoken/")
```
Comment (hohwille) [inline modules/security-csrf/src/main/java/com/devonfw/module/security/csrf/service/api/rest/CsrfRestService.java:26]: could be just `token` then...

File: modules/security-csrf/src/main/java/com/devonfw/module/security/csrf/service/impl/rest/CsrfRestServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.security.web.csrf.CsrfToken;
+import org.springframework.security.web.csrf.CsrfTokenRepository;
+
+import com.devonfw.module.security.csrf.service.api.rest.CsrfRestService;
+
+/**
+ * CsrfRestServiceImpl
```
Comment (hohwille) [inline modules/security-csrf/src/main/java/com/devonfw/module/security/csrf/service/impl/rest/CsrfRestServiceImpl.java:18]:
```
Implementation of {@link CsrfRestService}.
```
If we want some generic JavaDoc without further details.

File: modules/security-csrf/src/main/java/com/devonfw/module/security/csrf/service/api/rest/CsrfRestService.java [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.core.Context;
+
+import org.springframework.security.web.csrf.CsrfToken;
+
+import com.devonfw.module.rest.common.api.RestService;
+
+/**
+ * The CSRF security REST service provides access to the csrf token, the authenticated user's meta-data. Furthermore, it
```
Comment (hohwille) [inline modules/security-csrf/src/main/java/com/devonfw/module/security/csrf/service/api/rest/CsrfRestService.java:14]:
```
The CSRF REST service provides access to get the CSRF token used to protect against <a href="https://owasp.org/www-community/attacks/csrf">CSRF attacks</a>.
```

All other JavaDoc is wrong/obsolete.

File: templates/server/src/main/resources/archetype-resources/core/src/main/java/__packageInPathFormat__/general/service/impl/config/BaseWebSecurityConfig.java [hunk 13]
```diff
@@ -75,8 +75,8 @@ public void configure(HttpSecurity http) throws Exception {
         // define all urls that are not to be secured
         .authorizeRequests().antMatchers(unsecuredResources).permitAll().anyRequest().authenticated().and()
 
-        // activate crsf check for a selection of urls (but not for login & logout)
-        .csrf().requireCsrfProtectionMatcher(new CsrfRequestMatcher()).and()
+        //disabling CSRF
+        .csrf().disable()
```
Comment (hohwille) [inline templates/server/src/main/resources/archetype-resources/core/src/main/java/__packageInPathFormat__/general/service/impl/config/BaseWebSecurityConfig.java:79]: how does out spring-boot-starter enable it back?
Did you consider a property or to detect somehow if the starter is present?

File: documentation/guide-csrf.asciidoc [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+ public void configure(HttpSecurity http) throws Exception {
+ .....
+ http
+ .....
+	.csrf().requireCsrfProtectionMatcher(new CsrfRequestMatcher())
+ ....
+ if (this.corsEnabled) {
+    http.addFilterBefore(getCorsFilter(), CsrfFilter.class);
+  }
+ }
```
Comment (hohwille) [inline documentation/guide-csrf.asciidoc:45]: The idea of a starter would ideally be that the user is not required to do such changes manually. Just add or remove the starter should do the job if possible.

File: templates/server/src/main/resources/archetype-resources/core/src/main/java/__packageInPathFormat__/general/service/impl/config/BaseWebSecurityConfig.java [hunk 15]
```diff
@@ -88,10 +88,6 @@ public void configure(HttpSecurity http) throws Exception {
         // register login and logout filter that handles rest logins
         .addFilterAfter(getSimpleRestAuthenticationFilter(), BasicAuthenticationFilter.class)
         .addFilterAfter(getSimpleRestLogoutFilter(), LogoutFilter.class);
-
-    if (this.corsEnabled) {
-      http.addFilterBefore(getCorsFilter(), CsrfFilter.class);
```
Comment (hohwille) [inline templates/server/src/main/resources/archetype-resources/core/src/main/java/__packageInPathFormat__/general/service/impl/config/BaseWebSecurityConfig.java:93]: how will the starter add this filter? Could/should be done in `CsrfAutoConfiguration`.

**Review by hohwille — CHANGES_REQUESTED** [review]
@sujith-mn thanks for this PR. You are on the right track and most things look perfectly fine. 👍 
However, I have commented some places for rework.
Please have a look and try to update this PR accordingly.

**Comment by hohwille** [comment]
From travis log:
```
[ERROR]     'dependencies.dependency.version' for com.devonfw.java.starters:devon4j-starter-security:jar is missing. @ com.devonfw.java.starters:devon4j-starter-security-csrf:${devon4j.version}, /home/travis/build/devonfw/devon4j/starters/starter-security-csrf/pom.xml, line 18, column 17

[ERROR]     'dependencies.dependency.version' for com.devonfw.java.modules:devon4j-security-csrf:jar is missing. @ com.devonfw.java.starters:devon4j-starter-security-csrf:${devon4j.version}, /home/travis/build/devonfw/devon4j/starters/starter-security-csrf/pom.xml, line 23, column 17
```

Can you please merge my proposal #316 using git and not manually? You are missing the BOM changes (and maybe other things that can easily be forgotten by "copy & paste merging"). Any chance we can get this story into our release?

**Comment by hohwille** [comment]
Your code still fails - see travis logs:
```
[INFO] [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project basic-core: Compilation failure

[INFO] [ERROR] /home/travis/build/devonfw/devon4j/templates/server/target/test-classes/projects/basic/project/basic/core/src/main/java/it/pkg/general/service/impl/config/BaseWebSecurityConfig.java:[90,16] cannot find symbol

[INFO] [ERROR]   symbol: variable webSecurityConfigurer

[INFO] [ERROR] -> [Help 1]

[INFO] org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project basic-core: Compilation failure

[INFO] /home/travis/build/devonfw/devon4j/templates/server/target/test-classes/projects/basic/project/basic/core/src/main/java/it/pkg/general/service/impl/config/BaseWebSecurityConfig.java:[90,16] cannot find symbol

[INFO]   symbol: variable webSecurityConfigurer
```

**Comment by hohwille** [comment]
https://github.com/sujith-mn/devon4j/blob/dev_csrf/templates/server/src/main/resources/archetype-resources/core/src/main/java/__packageInPathFormat__/general/service/impl/config/BaseWebSecurityConfig.java#L90
But seems to miss declaring this variable somewhere here:
https://github.com/sujith-mn/devon4j/blob/dev_csrf/templates/server/src/main/resources/archetype-resources/core/src/main/java/__packageInPathFormat__/general/service/impl/config/BaseWebSecurityConfig.java#L43

Also it seems that in my PR #316 I did not include the archetype changes and I now figured out that I did do them in `target` and not in `src` - I was chased by other project work and did this on the run ;)

File: modules/security-csrf/src/main/java/com/devonfw/module/security/csrf/service/impl/rest/CsrfRestServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+  @PermitAll
+  public CsrfToken getCsrfToken(HttpServletRequest request, HttpServletResponse response) {
+
+    CsrfToken token = this.csrfTokenRepository.loadToken(request);
+    if (token == null) {
+      LOG.error("No CsrfToken could be found - instantiating a new Token");
+      token = this.csrfTokenRepository.generateToken(request);
+      this.csrfTokenRepository.saveToken(token, request, response);
+    }
+    return token;
```
Comment (hohwille) [inline modules/security-csrf/src/main/java/com/devonfw/module/security/csrf/service/impl/rest/CsrfRestServiceImpl.java:36]: you are not converting to `CsrfTokenImpl`. If this already works without any custom JSON mapping, why do we need `CsrfTokenImpl` then?

File: templates/server/src/main/resources/archetype-resources/core/src/main/java/__packageInPathFormat__/general/common/impl/config/CsrfTokenImpl.java [hunk 17]
```diff
@@ -57,4 +57,4 @@ public void setToken(String token) {
     this.token = token;
   }
 
-}
+}
```
Comment (hohwille) [inline templates/server/src/main/resources/archetype-resources/core/src/main/java/__packageInPathFormat__/general/common/impl/config/CsrfTokenImpl.java:60]: We can remove this class from app template if it is not used anymore.

File: templates/server/src/main/resources/archetype-resources/core/src/main/java/__packageInPathFormat__/general/service/impl/config/BaseWebSecurityConfig.java [hunk 18]
```diff
@@ -74,23 +77,23 @@ public void configure(HttpSecurity http) throws Exception {
...
         .logout().logoutSuccessUrl("/login.html").and()
-
         // register login and logout filter that handles rest logins
         .addFilterAfter(getSimpleRestAuthenticationFilter(), BasicAuthenticationFilter.class)
         .addFilterAfter(getSimpleRestLogoutFilter(), LogoutFilter.class);
 
+	//disable CSRF protection by default, use csrf starter to override.
+	http = http.csrf().disable();
+	//load starters as pluggins.
+    http = this.webSecurityConfigurer.configure(http);
```
Comment (hohwille) [inline templates/server/src/main/resources/archetype-resources/core/src/main/java/__packageInPathFormat__/general/service/impl/config/BaseWebSecurityConfig.java:93]: please move to the beginning of this method so the custom config code comes after and allows overriding by usere.

**Review by hohwille — CHANGES_REQUESTED** [review]
@sujith-mn thanks for your fixes and rework. This PR is almost ready to be merged 👍 
However, can you check my review comments (esp. about `CsrfTokenImpl`) and update this PR?
BTW: to speed up, I directly edited and improved your documentation that has been outdated due to the latest changes.

File: templates/server/src/main/resources/archetype-resources/core/src/main/java/__packageInPathFormat__/general/service/impl/config/BaseWebSecurityConfig.java (hunk 18 above)
Comment (hohwille) [reply templates/server/src/main/resources/archetype-resources/core/src/main/java/__packageInPathFormat__/general/service/impl/config/BaseWebSecurityConfig.java:93]: also fix indendation please

File: documentation/guide-csrf.asciidoc [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+    .....
+  }  
+}
+----
+
+Please note that the exact name (`@Named("CsrfRequestMatcher")`) is required here to ensure your custom implementation will be injected properly.
+
+== Configure devon4ng client for CSRF
+
+TODO: Link to devon4ng guide CSRF section.
```
Comment (hohwille) [inline documentation/guide-csrf.asciidoc:70]: I left this TODO so please add the deep-link to the devon4ng documentation here once ready.

Comment (sujith-mn) [reply documentation/guide-csrf.asciidoc:70]: sure.. link will be added later.

File: modules/security-csrf/src/main/java/com/devonfw/module/security/csrf/service/impl/rest/CsrfRestServiceImpl.java (hunk 16 above)
Comment (sujith-mn) [reply modules/security-csrf/src/main/java/com/devonfw/module/security/csrf/service/impl/rest/CsrfRestServiceImpl.java:36]: Thank you for editing..

**Review by hohwille — APPROVED** [review]
@sujith-mn thanks for completing this PR and all your efforts. 👍 
I hearby approve but want to point out the following open issues:
* [ ] As we had introduced `CsrfTokenImpl` once due to the fact that Jackson mapping of the default `CsrfToken` type from spring-security was not properly mapped to JSON by Jackson: Have you tested the CSRF rest service after you enabled CSRF? Can you provide the resulting example JSON via Copy&Paste so we can verify it is sane?
* [ ] we should not forget to resolve the TODO from the documentation before the release. However, I am fine with merging now and fixing later :)

**Comment by sujith-mn** [comment]
> @sujith-mn thanks for completing this PR and all your efforts. 👍
> I hearby approve but want to point out the following open issues:
> 
> * [ ]  As we had introduced `CsrfTokenImpl` once due to the fact that Jackson mapping of the default `CsrfToken` type from spring-security was not properly mapped to JSON by Jackson: Have you tested the CSRF rest service after you enabled CSRF? Can you provide the resulting example JSON via Copy&Paste so we can verify it is sane?

Here is the JSON:
{
    "token": "f26f94f5-6d71-4cf3-b1ee-316f64967f74",
    "parameterName": "_csrf",
    "headerName": "X-CSRF-TOKEN"
}

> * [ ]  we should not forget to resolve the TODO from the documentation before the release. However, I am fine with merging now and fixing later :)

Yes.. documentation will be done before release.

**Pull request merged**